<a class="anchor" id="About_Data"></a>
# <p style="padding:10px;background-color:#f6cd04;margin:0;color:#0c4da3;font-family:newtimeroman;font-size:150%;text-align:center;border-radius: 20px 20px;overflow:hidden;font-weight:500">Police 🚨👮‍♀</p> 



<img src="https://media0.giphy.com/media/G825ArmDSn6Ue5g152/giphy.gif?cid=6c09b952m7vzpzgzjpmdnsq601p0jwqqiksohe8x8tzgbl0f&ep=v1_internal_gif_by_id&rid=giphy.gif&ct=g" width="100%" height="100%">


<a class="anchor" id="About_Data"></a>
# <p style="padding:10px;background-color:#f6cd04;margin:0;color:#0c4da3;font-family:newtimeroman;font-size:150%;text-align:center;border-radius: 15px 50px;overflow:hidden;font-weight:500">Introduction  🚓</p> 



<div style = 'border : 5px solid #0c4da3; background-color:lightred;padding:10px'>

This project uses `Python's Pandas library` to clean and analyze 
[Police traffic Dataset](https://www.kaggle.com/datasets/melihkanbay/police/data).

The goal is to `practice data analysis` by identifying and fixing issues like missing values and inconsistencies in the dataset.

<a class="anchor" id="Cleaning_Data"></a>
# <p style="padding:10px;background-color:#f6cd04;margin:0;color:#0c4da3;font-family:newtimeroman;font-size:150%;text-align:center;border-radius: 50px 15px;overflow:hidden;font-weight:500">Data Cleaning  🔍🛠</p> 



In [1]:
import pandas as pd
import warnings
warnings.simplefilter("ignore")

In [2]:
police = pd.read_csv("police.csv")

In [3]:
police.head().style.background_gradient(cmap='Blues')

In [4]:
police.shape

(91741, 15)

In [5]:
police.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           91741 non-null  object 
 1   stop_time           91741 non-null  object 
 2   county_name         0 non-null      float64
 3   driver_gender       86406 non-null  object 
 4   driver_age_raw      86414 non-null  float64
 5   driver_age          86120 non-null  float64
 6   driver_race         86408 non-null  object 
 7   violation_raw       86408 non-null  object 
 8   violation           86408 non-null  object 
 9   search_conducted    91741 non-null  bool   
 10  search_type         3196 non-null   object 
 11  stop_outcome        86408 non-null  object 
 12  is_arrested         86408 non-null  object 
 13  stop_duration       86408 non-null  object 
 14  drugs_related_stop  91741 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 9.3+

In [6]:
police.dtypes

stop_date              object
stop_time              object
county_name           float64
driver_gender          object
driver_age_raw        float64
driver_age            float64
driver_race            object
violation_raw          object
violation              object
search_conducted         bool
search_type            object
stop_outcome           object
is_arrested            object
stop_duration          object
drugs_related_stop       bool
dtype: object

In [7]:
police.isnull().sum()

stop_date                 0
stop_time                 0
county_name           91741
driver_gender          5335
driver_age_raw         5327
driver_age             5621
driver_race            5333
violation_raw          5333
violation              5333
search_conducted          0
search_type           88545
stop_outcome           5333
is_arrested            5333
stop_duration          5333
drugs_related_stop        0
dtype: int64

In [8]:
police.describe().style.background_gradient(cmap = "Blues")

In [9]:
police.describe(include='O').T

count unique                 top   freq
stop_date      91741   3768          2012-01-11     69
stop_time      91741   1436               11:00    323
driver_gender  86406      2                   M  62895
driver_race    86408      5               White  62158
violation_raw  86408     12            Speeding  48463
violation      86408      6            Speeding  48463
search_type     3196     24  Incident to Arrest   1219
stop_outcome   86408      6            Citation  77006
is_arrested    86408      2               False  83479
stop_duration  86408      5            0-15 Min  69543

<div style = 'border : 5px solid #0c4da3; background-color:lightred;padding:10px'>

#### <u><span style="color:#08306B;">  Observation: </span></u>
 - The police dataset consists of  `91741 rows` and `15 columns`.
 - It has all `NAN` values in `"county_name"` column.
 - There are some coloumns that have uncorrect data type.

In [10]:
police = police.drop(
    labels='county_name',
    axis=1)

In [11]:
police.head().style.background_gradient(cmap='Blues')

<a class="anchor" id="Duplicated"></a>
# <p style="padding:10px;background-color:#f6cd04;margin:0;color:#0c4da3;font-family:newtimeroman;font-size:100%;text-align:center;border-radius: 50px 50px;overflow:hidden;font-weight:200">I. Duplicated Data 🔁</p> 



In [12]:
police.duplicated().sum()

479

- #### <span style="color:#08306B;">  There are about `479` duplicated rows, So let's `drop` it.</span>

In [13]:
police.drop_duplicates(inplace = True)

In [14]:
police.head().style.background_gradient(cmap='Blues')

In [15]:
police.duplicated().sum()

0

<a class="anchor" id="Missing"></a>
# <p style="padding:10px;background-color:#f6cd04;margin:0;color:#0c4da3;font-family:newtimeroman;font-size:100%;text-align:center;border-radius: 50px 50px;overflow:hidden;font-weight:200">II. Missing Data ❔❔</p> 



In [16]:
def missing_values_table(df):
  mis_val = df.isnull().sum()
  mis_val_percent = 100*df.isnull().sum() / len(df)
  mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
  mis_val_table_ren_columns = mis_val_table.rename(columns = {0:'Missing Values' , 1:'of Total Values %'})
  return mis_val_table_ren_columns.round(2)

In [17]:
print("Missing In Police Data BEFORE :- ")
missing_values_table(police).style.background_gradient(cmap='Blues')

Missing In Police Data BEFORE :- 


- #### <span style="color:#08306B;"> From the previous table `'search_type'` Column has more than `50% null values`, So let's `drop` it. </span>

In [18]:
police.drop('search_type', axis=1 , inplace=True)

- #### <span style="color:#08306B;">  There are about `5123` Nan rows, So let's `drop` it.</span>

In [19]:
police.dropna(subset = ['driver_race'], inplace=True)

In [20]:
print("Missing In Police Data :- ")
missing_values_table(police).style.background_gradient(cmap='Blues')

Missing In Police Data :- 


- #### <span style="color:#08306B;">  `'driver_age'` and `'driver_gender'` columns also have null values, So how we can deal with them. </span>

In [21]:
police['driver_age'].unique()

array([20., 40., 33., 19., 21., 23., 36., 35., 26., 39., 50., 47., 52.,
       38., 41., 28., 46., 32., 58., 27., 43., 24., 17., 22., 44., 48.,
       34., 55., 51., 45., 29., 31., 37., 67., 25., 42., 30., 18., 57.,
       49., 61., 70., nan, 53., 60., 59., 64., 63., 54., 56., 66., 62.,
       69., 81., 68., 76., 75., 74., 82., 65., 72., 16., 77., 73., 71.,
       80., 79., 78., 88., 15., 83., 86., 84., 85., 87., 99., 89., 94.,
       90.])

In [22]:
mean_age = police['driver_age'].mean()

In [23]:
police['driver_age'] = police['driver_age'].fillna(mean_age)

In [24]:
police['driver_gender'].unique()

array(['M', 'F', nan], dtype=object)

In [25]:
police.dropna(subset = ['driver_gender','driver_age_raw'], inplace=True)


In [26]:
print("Missing In Police Data After :- ")
missing_values_table(police).style.background_gradient(cmap='Blues')

Missing In Police Data After :- 


- #### <span style="color:#08306B;">  Now our data is free of missing values `:)` </span>

In [27]:
police.head().style.background_gradient(cmap='Blues')

<a class="anchor" id="correction"></a>
# <p style="padding:10px;background-color:#f6cd04;margin:0;color:#0c4da3;font-family:newtimeroman;font-size:100%;text-align:center;border-radius: 50px 50px;overflow:hidden;font-weight:200">III. Data Correction ⚙️</p> 



In [28]:
police.insert(1,'stop_year',0)
police['stop_year'] = police['stop_date'].str[:4]
police['stop_year'] = police['stop_year'].astype('int64')

In [29]:
police.insert(2,'stop_month',0)
police['stop_month'] = police['stop_date'].str[5:7]

In [30]:
police.insert(4,'stop_hour',0)

In [31]:
police['stop_hour'] = police['stop_time'].str[:2]
police['stop_hour'] = police['stop_hour'].astype('int64')

In [32]:
police.insert(5,'stop_daytime','')

In [33]:
police.loc[(police['stop_hour'] >= 6) & (police['stop_hour'] <= 11),'stop_daytime'] = 'Morning'
police.loc[(police['stop_hour'] >= 12) & (police['stop_hour'] <= 17),'stop_daytime'] = 'Afternoon'
police.loc[(police['stop_hour'] >= 18) & (police['stop_hour'] <= 21),'stop_daytime'] = 'Evening'
police.loc[(police['stop_hour'] >= 22) | (police['stop_hour'] <= 5),'stop_daytime'] = 'Night'

In [34]:
police.head().style.background_gradient(cmap='Blues')

In [35]:
police = police.drop(labels=['stop_date','stop_time','driver_age_raw'],axis=1)


In [36]:
police.head().style.background_gradient(cmap='Blues')

In [37]:
police['driver_age'] = police['driver_age'].astype('int64')

In [38]:
police['stop_month'].unique()

array(['01', '02', '03', '04', '06', '07', '08', '09', '10', '11', '12',
       '05'], dtype=object)

In [39]:
police['stop_month'] = police['stop_month'].replace({'01':'Jan' ,'02':'Feb' ,'03':'Mar', '04':'Apr','05':'May','06':'June','07':'July','08':'Aug','09':'Sept','10':'Oct','11':'Nov','12':'Dec'})

In [40]:
police['driver_gender'] = police['driver_gender'].replace({'M':'Male' ,'F':'Female'})

In [42]:
police.head().style.background_gradient(cmap='Blues')

In [41]:
for column in police:
    print("Frequancy of "  + column + " column :")
    print(police[column].value_counts())
    print("-------------------------------------------\n")

Frequancy of stop_year column :
2012    10383
2006    10095
2007     8843
2014     8812
2015     8197
2008     8127
2011     7572
2013     7407
2009     7230
2010     6988
2005     2482
Name: stop_year, dtype: int64
-------------------------------------------

Frequancy of stop_month column :
Jan     7929
May     7463
Nov     7400
Mar     7295
Oct     7269
June    7160
Aug     7154
Apr     7094
July    7059
Sept    6987
Dec     6722
Feb     6604
Name: stop_month, dtype: int64
-------------------------------------------

Frequancy of stop_hour column :
10    6976
9     6442
11    5567
14    5339
23    5240
8     4930
0     4837
13    4601
15    4569
7     4359
1     4007
16    3598
12    3385
22    3211
6     3195
20    3156
17    2713
2     2710
19    2440
18    2040
21    1350
3     1067
4      216
5      188
Name: stop_hour, dtype: int64
-------------------------------------------

Frequancy of stop_daytime column :
Morning      31469
Afternoon    24205
Night        21476
Evening    

In [43]:
police = police[police['stop_outcome'] != 'N/D']
police['stop_outcome'].value_counts()

Citation            76776
Warning              5287
Arrest Driver        2538
No Action             588
Arrest Passenger      357
Name: stop_outcome, dtype: int64

In [44]:
police = police[police['stop_duration'] != '2']
police = police[police['stop_duration'] != '1']
police['stop_duration'].value_counts()

0-15 Min     68847
16-30 Min    13511
30+ Min       3186
Name: stop_duration, dtype: int64

In [45]:
police.head().style.background_gradient(cmap='Blues')

In [46]:
police.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 85544 entries, 0 to 91740
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   stop_year           85544 non-null  int64 
 1   stop_month          85544 non-null  object
 2   stop_hour           85544 non-null  int64 
 3   stop_daytime        85544 non-null  object
 4   driver_gender       85544 non-null  object
 5   driver_age          85544 non-null  int64 
 6   driver_race         85544 non-null  object
 7   violation_raw       85544 non-null  object
 8   violation           85544 non-null  object
 9   search_conducted    85544 non-null  bool  
 10  stop_outcome        85544 non-null  object
 11  is_arrested         85544 non-null  object
 12  stop_duration       85544 non-null  object
 13  drugs_related_stop  85544 non-null  bool  
dtypes: bool(2), int64(3), object(9)
memory usage: 8.6+ MB


<a id="Insights 👀"></a>  
### <p style="padding:5px;background-color:#f6cd04;margin:0;color:#0c4da3;font-family:newtimeroman;font-size:150%;text-align:center;border-radius: 50px 50px;overflow:hidden;font-weight:200">  Insights 💡</p>


<div style = 'border : 5px solid #0c4da3; background-color:lightred;padding:10px'>

### <u><span style="color:#08306B">Upon initial inspection, the following is observed:</span></u>

 -    We have now no missing & duplicated values.
    
 -   The majority of features are categorical which we have `9` features numerical`['stop_month', 'stop_daytime', 'driver_gender', 'driver_race', 'violation_raw','violation','stop_outcome','is_arrested','stop_duration']` and the other numerical.
    
 - We now have new columns also as `'stop_year','stop_month','stop_hour','stop_daytime'` that help to detect the time highly.
    
- Most of the data are `Males`, also the most are `not arrested` and `don't have drugs_related_stop`.
    
